core/corehttp/gateway_handler.go

```diff
@@ -296,36 +297,26 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 		return
 	}
 
-	dr, err := i.api.Unixfs().Get(r.Context(), resolvedPath)
-	if err != nil {
-		webError(w, "ipfs cat "+escapedURLPath, err, http.StatusNotFound)
+	// Finish early if client already has matching Etag
+	// (suffix match to cover both direct CID and DirIndex cases)
+	cidEtagSuffix := resolvedPath.Cid().String() + `"`
+	if strings.HasSuffix(r.Header.Get("If-None-Match"), cidEtagSuffix) {
```

Comment:
I switched this to exact match in  26c122efc063284bb67799e7ec3f78077ef73c87. Better safe than sorry.

Etag is now based on CID and response format, and in the future, we will also include CID of a selector, if present.
This is mostly to avoid issues with poorly written/configured HTTP caches/CDNs.